lib/src/mt_dhcp.c

```diff
@@ -0,0 +1,579 @@
+    *options++ = dhcp_impl->server_ip[1];
+    *options++ = dhcp_impl->server_ip[2];
+    *options++ = dhcp_impl->server_ip[3];
+  }
+  mt_pthread_mutex_unlock(&dhcp_impl->mutex);
+  *options++ = DHCP_OPTION_PARAMETER_REQUEST_LIST;
+  *options++ = 3;
+  *options++ = DHCP_OPTION_SUBNET_MASK;
+  *options++ = DHCP_OPTION_ROUTER;
+  *options++ = DHCP_OPTION_DNS_SERVER;
+  *options++ = DHCP_OPTION_END;
+  hdr_offset += sizeof(*dhcp) + RTE_PTR_DIFF(options, dhcp->options);
+
+  mt_mbuf_init_ipv4(pkt);
+  pkt->pkt_len = pkt->data_len = hdr_offset;
+
+  /* update length */
+  ip->total_length = htons(pkt->pkt_len - sizeof(*eth));
+  udp->dgram_len = htons(pkt->pkt_len - sizeof(*eth) - sizeof(*ip));
+
+  /* send dhcp request packet */
+  uint16_t send = mt_dev_tx_sys_queue_burst(impl, port, &pkt, 1);
+  if (send < 1) {
+    err_once("%s(%d), tx fail\n", __func__, port);
+    rte_pktmbuf_free(pkt);
+    return -EIO;
+  }
+
+  mt_pthread_mutex_lock(&dhcp_impl->mutex);
+  dhcp_impl->status = MT_DHCP_STATUS_REQUESTING;
+  mt_pthread_mutex_unlock(&dhcp_impl->mutex);
+
+  info("%s(%d), dhcp request sent\n", __func__, port);
+
+  return 0;
+}
+
+static int dhcp_recv_offer(struct mtl_main_impl* impl, struct mt_dhcp_hdr* offer,
+                           enum mtl_port port) {
+  struct mt_dhcp_impl* dhcp_impl = get_dhcp(impl, port);
+  mt_pthread_mutex_lock(&dhcp_impl->mutex);
+  if (dhcp_impl->status != MT_DHCP_STATUS_DISCOVERING) {
+    dbg("%s(%d), not in discovering status\n", __func__, port);
+    mt_pthread_mutex_unlock(&dhcp_impl->mutex);
+    return -EIO;
+  }
+  rte_memcpy(dhcp_impl->ip, &offer->yiaddr, MTL_IP_ADDR_LEN);
+  mt_pthread_mutex_unlock(&dhcp_impl->mutex);
+  info("%s(%d), received dhcp offer\n", __func__, port);
+  info("%s(%d), ip address: %s\n", __func__, port,
+       inet_ntoa(*(struct in_addr*)(char*)&offer->yiaddr));
+  uint8_t* options = offer->options;
+  while (*options != DHCP_OPTION_END) {
+    if (*options == DHCP_OPTION_SUBNET_MASK)
+      dbg("%s(%d), subnet mask: %s\n", __func__, port,
+          inet_ntoa(*(struct in_addr*)(options + 2)));
+    if (*options == DHCP_OPTION_ROUTER)
+      dbg("%s(%d), default gateway: %s\n", __func__, port,
+          inet_ntoa(*(struct in_addr*)(options + 2)));
+    if (*options == DHCP_OPTION_DNS_SERVER) {
+      for (int i = 0; i < options[1] / 4; ++i)
+        dbg("%s(%d), dns server %d: %s\n", __func__, port, i,
+            inet_ntoa(*(struct in_addr*)(options + 2 + i * 4)));
+    }
+    if (*options == DHCP_OPTION_SERVER_IDENTIFIER)
+      rte_memcpy(dhcp_impl->server_ip, &options[2], MTL_IP_ADDR_LEN);
+    options += options[1] + 2;
+  }
+
+  dhcp_send_request(impl, port);
+
+  return 0;
+}
+
+/* renew at t1 after ack */
+static void dhcp_renew_handler(void* param) {
+  struct mt_dhcp_impl* dhcp_impl = param;
+  mt_pthread_mutex_lock(&dhcp_impl->mutex);
+  dhcp_impl->status = MT_DHCP_STATUS_RENEWING;
+  mt_pthread_mutex_unlock(&dhcp_impl->mutex);
```

Comment:
I see many places do similar things:
```
  mt_pthread_mutex_lock(&dhcp_impl->mutex);
  dhcp_impl->status = xxxx;
  mt_pthread_mutex_unlock(&dhcp_impl->mutex);
```
Can we declare it as static inline function:
```
static inline void dhcp_set_status(dhcp, status) {
  lock;
  set status;
  unlock;
}
```